js/classes/BaseCalendar.js

```diff
@@ -0,0 +1,579 @@
+     * @param {number} year
+     * @param {number} month
+     * @param {number} day
+     * @return {Boolean}
+     */
+    _showDay(year, month, day)
+    {
+        return showDay(year, month, day, this._preferences);
+    }
+
+    /*
+     * Will check if the inputs for today are all filled and then enable the button, if not.
+     */
+    _checkTodayPunchButton()
+    {
+        const today = new Date();
+        const isCurrentMonth = (today.getMonth() === this._calendarDate.getMonth() && today.getFullYear() === this._calendarDate.getFullYear());
+        let enableButton = false;
+        if (isCurrentMonth)
+        {
+            const dateKey = generateKey(today.getFullYear(), today.getMonth(), today.getDate());
+            const inputs = $('#' + dateKey + ' input[type="time"]');
+            let allInputsFilled = true;
+            for (let input of inputs)
+            {
+                allInputsFilled &= $(input).val().length !== 0;
+            }
+            enableButton = !allInputsFilled;
+        }
+        this._togglePunchButton(enableButton);
+    }
+
+    /**
+     * Adds the next missing entry on the actual day and updates calendar.
+     */
+    punchDate()
+    {
+        const now = new Date(),
+            year = now.getFullYear(),
+            month = now.getMonth(),
+            day = now.getDate(),
+            hour = now.getHours(),
+            min = now.getMinutes();
+
+        if (this._getCalendarMonth() !== month ||
+            this._getCalendarYear() !== year ||
+            !this._showDay(year, month, day))
+        {
+            return;
+        }
+
+        const value = hourMinToHourFormatted(hour, min);
+        const key = generateKey(year, month, day);
+        const inputs = $('#' + key + ' input[type="time"]');
+        for (const element of inputs)
+        {
+            if ($(element).val().length === 0)
+            {
+                $(element).val(value);
+                this._updateTimeDayCallback(key);
+                break;
+            }
+        }
+    }
+
+    /**
+     * Based on the key of the input, updates the values for total in DB and display it on page.
+     * @param {string} key
+     */
+    _updateTimeDayCallback(key)
+    {
+        this._updateTimeDay(key);
+        this._updateLeaveBy();
+        this._updateBalance();
+    }
+
+    /**
+     * Based on the dateKey of the input, if an array of valid time values is given, the internal DB values are replaced.
+     * If they're invalid, the internal value is removed.
+     * @param {string} dateKey
+     * @param {Array.string} newValues Time values
+     */
+    _updateDbEntry(dateKey, newValues)
+    {
+        let validatedTimes = this._validateTimes(newValues);
+        if (validatedTimes.length > 0)
+        {
+            this._setStore(dateKey, validatedTimes);
+        }
+        else
+        {
+            this._removeStore(dateKey);
+        }
+    }
+
+    _updateDayTotal(key)
```

Comment:
Since you are adding documentation for all functions, maybe add to this one too?

```suggestion

    /**
     * Updates Day Total for a given day
     * @param {string} dateKey
     */
    _updateDayTotal(key)
```